src/docfx/lib/schema/JsonSchemaTransformer.cs

```diff
@@ -63,70 +63,91 @@ public JsonSchemaTransformer(JsonSchema schema)
                             xrefPropertiesGroupByUid[uid].Item2[key] = new Lazy<JToken>(
                             () =>
                             {
-                                // todo: change transform to `return` model instead of `replace` model
-                                var clonedObj = JsonUtility.DeepClone(obj);
-                                Transform(file, context, propertySchema, clonedObj[key], errors);
-                                return clonedObj[key];
+                                return Transform(file, context, propertySchema, value, errors);
                             }, LazyThreadSafetyMode.PublicationOnly);
                         }
                     }
-                    return schema.XrefProperties;
+                    return (schema.XrefProperties, node);
                 }
 
-                return default;
+                return (default, node);
             });
 
             return (errors, xrefPropertiesGroupByUid);
         }
 
-        private void Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
+        private JToken Transform(Document file, Context context, JsonSchema schema, JToken token, List<Error> errors)
         {
-            Traverse(schema, token, (subSchema, node) =>
+            return Traverse(schema, token, (subSchema, node) =>
             {
                 if (node.Type == JTokenType.Array || node.Type == JTokenType.Object)
                 {
                     // don't support array/object transform now
-                    return default;
+                    return (default, node);
                 }
 
-                node.Replace(TransformScalar(subSchema, file, context, node as JValue, errors));
-
-                return default;
+                var transformedScalar = TransformScalar(subSchema, file, context, node as JValue, errors);
+                JsonUtility.SetSourceInfo(transformedScalar, JsonUtility.GetSourceInfo(node));
+                return (default, transformedScalar);
             });
         }
 
-        private void Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, string[]> transform)
+        private JToken Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, (string[], JToken)> transform)
         {
             schema = _definitions.GetDefinition(schema);
+            var sourceInfo = JsonUtility.GetSourceInfo(token);
             if (schema == null)
             {
-                return;
+                return token;
             }
 
-            var transformedKeys = transform(schema, token) ?? Array.Empty<string>();
+            var (transformedKeys, transformedToken) = transform(schema, token);
+            transformedKeys = transformedKeys ?? Array.Empty<string>();
```

Comment:
can compare `token` and `transformedToken` for reference equality to detect if the token has been processed. If the token is processed, skip the entire switch block in line 107.
With this, the callback does not need to return `transformedKeys`

Reply:
when the token is process, if the token is JObject, maybe only a few properties are process, so reference equality is not working here